sympy/core/expr.py

```diff
@@ -4137,10 +4138,140 @@ def search_element(self, elem):
         return None
 
 
+def _power_gens(expr):
+    from sympy.ntheory.factor_ import perfect_power
+    # helper for generators to analyze the powers of the
+    # factors in expr
+    p = expr
+    if expr.is_Pow:
+        cb, e = expr.as_base_exp()
+        b = cb.as_coeff_Mul()[1] if not cb.is_Rational else cb
+        c, e = e.as_coeff_Mul(rational=True)
+        assert not e.is_Add
+        e /= c.q
+        # Poly differentiates between x and 1/x; to
+        # do so here, uncomment the 2 lines below
+        #if c.p < 0:
+        #    e = -e
+        if b.is_Rational and not e.is_Integer:
+            if b.is_Integer:
+                pp = perfect_power(b)
+                if pp:
+                    b, _exp = pp
+            else:
+                npp = perfect_power(b.p)
+                if npp:
+                    n, _exp = npp
+                    dpp = perfect_power(b.q, [_exp])
+                    if dpp:
+                        d, _ = dpp
+                        b = Rational(n, d, gcd=1)
+        p = Pow(b, e, evaluate=False) if e - 1 else b
+    return {p}
+
+
+def generators(self, all=False):
+    """return a set of expressions which would appear in the expanded
+    version of self as 1) the bases of powers with Integer exponents
+    or else as 2) powers with non-Integer exponents (but with numerator
+    free of leading Integer). Only expressions with free symbols are
+    returned.  If ``all=True`` then the set will include free symbols
+    of the generators, too.
+
+    EXAMPLES
+    ========
+
+    >>> from sympy.core.expr import generators
+    >>> from sympy.abc import x, y
+    >>> from sympy import Function, exp, cos
+    >>> f = Function('f')
+    >>> generators(1 + x + f(x))
+    {x, f(x)}
+
+    If an expression is a number, even symbolically, it will have no
+    generators:
+
+    >>> from sympy import Integral
+    >>> generators(Integral(f(x), (f(x), 1, 2)))
+    set()
+
+    If a Symbol does not appear independent of a generator, it will
+    not be included:
+
+    >>> generators(cos(1 + f(x)))
+    {cos(f(x) + 1)}
+
+    But generators embedded within defined functions can be included
+    by using ``all=True``:
+
+    >>> generators(cos(1 + f(x)), all=True)
+    {f(x), cos(f(x) + 1)}
+
+    When powers or exponential functions are returned, the leading
+    Integer of the numerator of the exponent/argument will not be
+    included
+    >>> generators(exp(2*x) + y**(-x/2) + y**(2*x))
+    {y**(x/2), y**x, exp(x)}
+    >>> generators(x/y**2)
```

Comment:
maybe generators should be split between those with positive coefficients and those with negative so this could be `({x},{y})`...and these doctests need to be written without sets.